fix: add connection state for net stream and gracefully handle failure

AkMo3 opened the pull request:
## What was wrong?

The Python NetStream class was too simple compared to its Go counterpart. It lacked proper state management and cleanup mechanisms, which meant:

- No way to track if a stream was open, closed for reading/writing, or reset
- No proper cleanup when streams closed (removing from SwarmConn, triggering ClosedStream events)
- No state validation before operations (could attempt to write to write-closed streams)
- Missing notification system for stream lifecycle events

Fixes: #300

## How was it fixed?

Enhanced NetStream with state management:
1. State Tracking: Added StreamState enum (OPEN, CLOSE_READ, CLOSE_WRITE, CLOSE_BOTH, RESET) with Trio-compatible async locks
2. Operation Validation: Added state checks before read/write/close operations to prevent invalid actions and provide clear error messages
3. Proper Cleanup: Implemented _remove() method that:
   1. Removes stream from swarm connection
   2. Notifies all listeners via ClosedStream events
   3. Handles reference counting
4. State Transitions: Automatic state updates based on:
    - EOF on read → close read side
    - Explicit close → close write side
    - Reset → close both sides
    - Triggers cleanup when fully closed

### To-Do

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/libp2p/py-libp2p/blob/main/newsfragments/README.md)

seetadev commented:
@AkMo3 : Great progress — appreciate the detailed improvements to `NetStream`! Thanks so much for this thoughtful and much-needed update.

You've addressed some of the core gaps in the original NetStream implementation with great care and precision. 

The new state management approach is a solid upgrade. Introducing the `StreamState` enum and enforcing operation validation brings much-needed structure and safety to stream handling. The `_remove()` logic and lifecycle event triggers are also spot-on — this makes the stream lifecycle far more reliable and observable.

Appreciate the structured cleanup flow via _remove(), which not only handles de-registration and event notifications but also ensures consistency with reference counting. Your attention to lifecycle transitions (e.g., EOF → close read side, reset → full close) is excellent and exactly what's been missing.

This PR indeed brings the Python libp2p implementation much closer in quality and behavior to other core libp2p modules.

Looking forward to the CI/CD fix and final cleanup — let me know if you'd like another pair of eyes on the failing jobs. Great work here, thank you :)

AkMo3 commented:
@seetadev Thankyou for the careful review! I have done the CI/CD fixes and final cleanup, you can run CI/CD jobs agains now.

seetadev commented:
> @seetadev Thankyou for the careful review! I have done the CI/CD fixes and final cleanup, you can run CI/CD jobs agains now.

@AkMo3 :  Thank you so much for the quick turnaround and for addressing the CI/CD issues — much appreciated! I’ve re-triggered the CI/CD jobs, and everything looks good so far.

As a final step before we wrap up this PR, could you please add the following:

1. **Newsfragment File:**
   A brief summary of this change under the appropriate category (feature, bugfix, etc.) using the standard newsfragment format. This will help us maintain a clean and automated changelog for upcoming releases.

2. **Developer Documentation:**
  Wish if you could explain what this PR does, how to use or extend it, and any dependencies or limitations. This is particularly helpful for future contributors or maintainers unfamiliar with this part of the codebase.

3. **Demo / Screencast (Optional but Valuable):**
   If you’re able to record a short screencast or gif showing the feature or component in action, that would be fantastic. It gives reviewers, testers, and future users a quick and intuitive understanding of your work.

Once those are in, I believe we’ll be in a great position to merge. You've done an excellent job here — thanks again for your thoroughness and dedication to quality.

Looking forward to the final polish. 

CCing @pacrob for his thoughts/feedback on the PR.

AkMo3 commented:
Fully agree on the documentation and demo part. I have added class level documentation regarding internal workings of the state, as you can see [here](https://py-libp2p--658.org.readthedocs.build/en/658/libp2p.network.stream.html#state-machine). 

Also, to serve as a demo, I have added a new example based upon the existing echo server, and provided the link to it as part of class documentation. Updated the news fragment as well.

AkMo3 commented on Makefile:
```diff
@@ -38,7 +38,7 @@ lint:
 
 test:
-	python -m pytest tests
+	python -m pytest -n auto tests
```
This actually enables tests to run in parallel, reducing the testing time multifold.

AkMo3 reviewed (commented):
Note:

AkMo3 reviewed (commented):
A few more improvements:

AkMo3 commented on tox.ini:
```diff
@@ -19,10 +19,10 @@ max_issue_threshold=1
-    core: pytest {posargs:tests/core}
-    interop: pytest {posargs:tests/interop}
+    core: pytest -n auto {posargs:tests/core}
+    interop: pytest -n auto {posargs:tests/interop}
     docs: make check-docs-ci
-    demos: pytest {posargs:tests/core/examples/test_examples.py}
+    demos: pytest -n auto {posargs:tests/core/examples/test_examples.py}
```
Reduces the CI time by half:
<img width="881" alt="image" src="https://github.com/user-attachments/assets/a18439fa-e0cd-4512-a9ff-077536cac482" />
